crates/lox-derive/src/lib.rs

```diff
@@ -0,0 +1,724 @@
+                                        next_line,
+                                    )?;
+
+                                    if #condition_shortcut line_matches {
+                                        #field_type_new::deserialize(lines)?
+                                    } else {
+                                        Err(crate::ndm::kvn::KvnDeserializerErr::<String>::UnexpectedKeyword {
+                                            found: next_line.to_string(),
+                                            expected: #expected_kvn_name.to_string(),
+                                        })?
+                                    }
+                                }
+                            }
+                        }
+                    }
+                };
+
+                Ok(quote! {
+                    #field_name: #parser,
+                })
+             })
+             .collect();
+
+        if let Err(e) = other_field_deserializers {
+            return e;
+        }
+
+        let other_field_deserializers = other_field_deserializers.unwrap();
+
+        quote! {
+            Ok(#type_name {
+                #(#other_field_deserializers)*
+            })
+        }
+    }
+}
+
+fn deserializers_for_struct_with_unnamed_fields(
+    type_name: &proc_macro2::Ident,
+    fields: &syn::punctuated::Punctuated<Field, syn::token::Comma>,
+) -> proc_macro2::TokenStream {
+    let field = fields
+        .first()
+        .expect("We expect exactly one item in structs with unnamed fields");
+
+    if &type_name.to_string() == "EpochType" {
+        return quote! {
+            Ok(#type_name (
+                crate::ndm::kvn::parser::parse_kvn_datetime_line_new(
+                    lines.next().unwrap()
+                ).map_err(|x| crate::ndm::kvn::KvnDeserializerErr::from(x))
+                .map(|x| x.1)?.full_value
+            ))
+        };
+    }
+
+    // Unwrap is okay because we expect this span to come from the source code
+    let field_type = extract_type_path(&field.ty)
+        .unwrap()
+        .span()
+        .source_text()
+        .unwrap();
+    let field_type_new = extract_type_path(&field.ty).unwrap();
+
+    let deserializer_for_kvn_type =
+        generate_call_to_deserializer_for_kvn_type_new(&field_type, field_type_new);
+
+    let deserializer_for_kvn_type = match deserializer_for_kvn_type {
+        Ok(deserializer_for_kvn_type) => deserializer_for_kvn_type,
+        Err(e) => return e.into(),
+    };
+
+    quote! {
+        Ok(#type_name (
+            #deserializer_for_kvn_type.value
+        ))
+    }
+}
+
+#[proc_macro_derive(KvnDeserialize, attributes(kvn))]
+pub fn derive_kvn_deserialize(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let item = syn::parse_macro_input!(item as syn::DeriveInput);
+    let type_name = &item.ident;
+    let is_value_unit_struct = is_value_unit_struct(&item);
+
+    let syn::Data::Struct(strukt) = item.data else {
```

Comment:
I've never seen and else blocked used with matching outside the context of an `if let` expression. Nice tip, thanks!

Reply:
It was my first time using it too.